tst/unit/test_container_iterator.cpp

```diff
@@ -137,3 +138,156 @@ TEST_CASE("Can pull variables from containers based on Metadata", "[ContainerIte
     }
   }
 }
+//Test wrapper to run a function multiple times
+template<typename InitFunc,typename PerfFunc>
+double performance_test_wrapper(const int n_burn, const int n_perf,
+    InitFunc init_func,PerfFunc perf_func){
+
+  //Initialize the timer and test
+  Kokkos::Timer timer;
+  init_func();
+
+  for( int i_run = 0; i_run < n_burn + n_perf; i_run++){
+    if(i_run == n_burn){
+      //Burn in time is over, start timing
+      Kokkos::fence();
+      timer.reset();
+    }
+
+    //Run the function timing performance
+    perf_func();
+  }
+
+  //Time it
+  Kokkos::fence();
+  double perf_time = timer.seconds();
+
+  //FIXME?
+  //Validate results?
+
+  return perf_time;
+
+}
+
+
+TEST_CASE("Container Iterator Performance",
+          "[ContainerIterator][performance]") {
+
+
+  const int N = 16; //Dimensions of blocks
+  const int n_burn = 500; //Num times to burn in before timing
+  const int n_perf = 500; //Num times to run while timing
+
+  //Make a raw ParArray4D for closest to bare metal looping
+  ParArray4D<Real> raw_array("raw_array",10,N,N,N);
+
+  //Make a function for initializing the raw ParArray4D
+  auto init_raw_array = [&]() {
+    par_for("Initialize ", DevSpace(),
+      0, 10-1,
+      0, N-1,
+      0, N-1,
+      0, N-1,
+      KOKKOS_LAMBDA(const int l, const int k, const int j, const int i) {
+        raw_array(l,k,j,i) = static_cast<Real>( (l+1)*(k+1)*(j+1)*(i+1) );
+      });
+  };
+
+  //Test performance iterating over variables (we should aim for this performance)
+  double time_raw_array = performance_test_wrapper( n_burn, n_perf,init_raw_array,
+    [&](){
+      par_for("Raw Array Perf", DevSpace(),
+        0, 10-1,
+        0, N-1,
+        0, N-1,
+        0, N-1,
+        KOKKOS_LAMBDA(const int l, const int k, const int j, const int i) {
+          raw_array(l,k,j,i) *= raw_array(l,k,j,i); //Do something trivial, square each term
+        });
+    });
+
+
+  //Make a container for testing performance
+  Container<Real> container;
+  Metadata m_in({Metadata::Independent});
+  Metadata m_out;
+  std::vector<int> scalar_block_size {N,N,N};
+  std::vector<int> vector_block_size {3,N,N,N};
```

Comment:
This is backwards, I think.  Should be
```suggestion
  std::vector<int> vector_block_size {N,N,N,3};
```

Reply:
I noticed this in the original test. Are we still using `LayoutRight` by default, and are we using these vectors like `{X,k,j,i}` and vectorizing across the last dimension? Then `{3,N,N,N}` is the best layout, so that consecutive elements in the vector are strided in memory. It might even make sense to use `{N,N,3,N}` so that the last dimension still vectorizes but consecutive elements are in the same page.

Reply:
The `3` should not be used in the fastest moving direction, which needs to vectorize.

Reply:
This line doesn't control the layout.  There's still some inconsistency (which we should probably eliminate) with respect to how we declare the size of different dimensions.  When adding a variable to a container, the vector of sizes passed in is interpreted like
```c++
vector_block_size {dim1, dim2, dim3, dim4, ...}
```
This translates into a ParArrayND like
```c++
ParArrayND<Real> data("name",dim4,dim3,dim2,dim1);
```
and the layout is ultimately controlled by Kokkos, I think.  So, again, with the current code, that line needs to change.  Leaving as is produces memory errors.

Reply:
I see, that makes much more sense. It should be `{N,N,N,3}` then if the dimensions get flipped